### PR TITLE
Add the childrenOf query filter for search results to the allowed filters.

### PR DESCRIPTION
**Pull request by cuyt**
* Allow the 'childrenOf' filter in the allowed set of filters
* Add appropriate type checking for this filter
* Add tests for search result filtering behaviour
* Fix some type assertions in the original results filter lists.

### CODE REVIEW

**Comment by cuyt** [comment]
Hello @marcoooo, 
On our project we've started to make use of this excellent wrapper library around the OLS API. I noticed a gap in the allowed filters on search results, namely the 'chlidrenOf' filter (as described on https://www.ebi.ac.uk/ols/docs/api), which I'd like to make use of. 
This PR simple lets the filter pass the results filter checks. It includes basic unit tests for the filter on one of the ontologies (Phi) in the OLS-docker image. 
I've found a couple of instances where I think the filter type check has a typo. One of the commits is a presumed fix for this.
This is my first time make a PR on github, so please let me know if anything is missing or should be done differently. 
I would greatly appreciate if you have a chance to look at the PR and if satisfactory, merge into the main repo.

Best regards,
Thomas

**Comment by marcoooo** [comment]
Hi Thomas, thanks for this, we'll have a look to your PR soon.

File: ebi/ols/api/base.py [hunk 1]
```diff
@@ -111,12 +111,19 @@ def filters_response(filters):
...
+                assertion_set = set(filters['childrenOf'].split(','))
+            elif type(filters['childrenOf']) is set:
+                assertion_set = filters['childrenOf']
+            else:
+                raise AssertionError("Wrong filter childrenOf %s" % filters['childrenOf'])
         if 'fieldList' in filters:
             if type(filters['fieldList']) is str:
                 assertion_set = set(filters['fieldList'].split(','))
-            elif type(filters['fieldList'] is set):
+            elif type(filters['fieldList']) is set:
```
Comment (marcoooo) [inline ebi/ols/api/base.py:126]: Well spotted! :-)

File: tests/test_basic.py [hunk 2]
```diff
@@ -229,6 +229,92 @@ def test_search_filters(self):
...
+        actual_terms = self.client.search(
+            query=query, 
+            ontology=ontology,
+            filters=filters)
+        actual_iris = { term.iri for term in actual_terms }
+        expected_iris = {
+                'http://purl.obolibrary.org/obo/PHI_1000003',
+                'http://purl.obolibrary.org/obo/PHI_1000002'
+        }
+        assert(actual_iris == expected_iris)
```
Comment (radome) [inline tests/test_basic.py:275]:
```suggestion
        self.assertEqual(actual_iris, expected_iris)
```

File: tests/test_basic.py [hunk 3]
```diff
@@ -229,6 +229,92 @@ def test_search_filters(self):
...
+            query=query, 
+            ontology=ontology,
+            filters=filters)
+        actual_iris = { term.iri for term in actual_terms }
+        expected_iris = {
+                'http://purl.obolibrary.org/obo/PHI_1000003',
+                'http://purl.obolibrary.org/obo/PHI_1000002',
+                'http://purl.obolibrary.org/obo/PHI_2000004'
+        }
+        assert(actual_iris == expected_iris)
```
Comment (radome) [inline tests/test_basic.py:299]:
```suggestion
        self.assertEqual(actual_iris, expected_iris)
```

File: tests/test_basic.py [hunk 4]
```diff
@@ -229,6 +229,92 @@ def test_search_filters(self):
...
+            'childrenOf': 'http://purl.obolibrary.org/obo/PHI_3000000',
+        }
+        # When
+        actual_terms = self.client.search(
+            query=query, 
+            ontology=ontology,
+            filters=filters)
+        actual_iris = { term.iri for term in actual_terms }
+        expected_iris = set()
+        assert(actual_iris == expected_iris)
```
Comment (radome) [inline tests/test_basic.py:316]:
```suggestion
        self.assertEqual(actual_iris, expected_iris)
```

**Review by radome — CHANGES_REQUESTED** [review]
Hi @cuyt, thank you for you PR.

I've added some small suggestions in order for the test suite to be consistent.

@marcoooo in general IMO we should prefer `isinstance` over `type` (see this interesting [post](https://switowski.com/blog/type-vs-isinstance)). Also we should probably review the use of `assert` and `AssertionError` in the code. I'll work on a PR about that and updating the `README` as well.

**Comment by cuyt** [comment]
Thanks for the quick review @radome! I've applied your suggestions and updated the `filter_response` `type` testing to `isinstance` per your suggestion.

On a different note: I've seen that the API also has the `api/select` endpoint which does some different kind of search (I'm not sure how it differs exactly from the `search` endpoint). At this point I don't think the ols-client wraps the `select` endpoint. Do you know if this is planned for the future?

Cheers,
Thomas

**Comment by marcoooo** [comment]
> Thanks for the quick review @radome! I've applied your suggestions and updated the `filter_response` `type` testing to `isinstance` per your suggestion.
> 
> On a different note: I've seen that the API also has the `api/select` endpoint which does some different kind of search (I'm not sure how it differs exactly from the `search` endpoint). At this point I don't think the ols-client wraps the `select` endpoint. Do you know if this is planned for the future?
> 
> Cheers,
> Thomas

We'll be happy to have a PR coming in proposing this integration :-), I am afraid we dedicated our first effort to wrapp the most needed services for our Ensembl requirements. We may integrate it in the future, but I can't tell you exactly when.

**Review by radome — APPROVED** [review]

**Pull request merged**